File: test/unit_controller.py
import os
import sys
import unittest

PYMOD_PATH = os.environ['PYMOD_PATH']
sys.path.insert(0, PYMOD_PATH)

import pyptzf


class TestControllerConfig(unittest.TestCase):

    def test_operators(self):
        # operator __eq__ and __ne__
        device = "/dev/foo"
        min = pyptzf.Position(0,0,0,0)
        max = pyptzf.Position(1,1,1,1)
        do_connect = False
        c1 = pyptzf.Controller.Config(device, min, max, do_connect)
        c2 = pyptzf.Controller.Config(device, min, max, do_connect)
        c3 = pyptzf.Controller.Config(device, min, max, True)
        self.assertEqual(c1, c2)
        self.assertNotEqual(c1, c3)

    def test_is_valid_position(self):
        device = "/dev/foo"
        min = pyptzf.Position(0,0,0,0)
        max = pyptzf.Position(1,1,1,1)
        do_connect = False

        c = pyptzf.Controller.Config(device, min, max, do_connect)
        p = pyptzf.Position()  # 0,0,0,0

        self.assertTrue(c.is_valid_position(min))
        self.assertTrue(c.is_valid_position(max))
        self.assertTrue(c.is_valid_position(p))

        offsets = (
            pyptzf.Position(1,0,0,0),
            pyptzf.Position(0,1,0,0),
            pyptzf.Position(0,0,1,0),
            pyptzf.Position(0,0,0,1),
        )

        for offset in offsets:
            self.assertFalse(c.is_valid_position(min - offset))
            self.assertFalse(c.is_valid_position(max + offset))


if __name__ == "__main__":
    unittest.main()

File: test/unit_controller.cpp
#include <sstream>
#include "controller.h"
#include "gtest/gtest.h"

#include "position.h"

namespace ptzf {
namespace {

// The fixture for testing class Controller.
class ControllerConfigTest : public ::testing::Test {};

// ensure Json (de)serialization works
TEST_F(ControllerConfigTest, TestJson) {
  Controller::Config c;
  EXPECT_EQ(c, Controller::Config::from_json(c.to_json()));
}

TEST_F(ControllerConfigTest, TestOperators) {
  // operator ==
  const std::string device("/dev/foo");
  const Position min{-1,-1,-1,-1};
  const Position max{1,1,1,1};
  const bool do_connect = false;
  Controller::Config c{device, min, max, do_connect};
  Controller::Config c2{device, min, max, do_connect};
  Controller::Config c3("/dev/bar", min, max, do_connect);
  EXPECT_EQ(c, c2);
  EXPECT_NE(c, c3);
}

TEST_F(ControllerConfigTest, TestIsValidPosition) {
  // operator ==
  const std::string device("/dev/foo");
  const Position min{-1,-1,-1,-1};
  const Position max{1,1,1,1};
  const bool do_connect = false;
  Controller::Config c{device, min, max, do_connect};

  EXPECT_TRUE(c.is_valid_position(min));
  EXPECT_TRUE(c.is_valid_position(max));
  EXPECT_TRUE(c.is_valid_position(Position()));

  Position x1{1,0,0,0};
  Position y1{0,1,0,0};
  Position z1{0,0,1,0};
  Position f1{0,0,0,1};

  for (const auto offset : {x1, y1, z1, f1}) {
    EXPECT_FALSE(c.is_valid_position(min - offset));
    EXPECT_FALSE(c.is_valid_position(max + offset));
  }
}

}  // namespace
}  // namespace ptzf

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
